calico/felix/ipsets.py

```diff
@@ -159,103 +176,137 @@ def on_tags_update(self, profile_id, tags):
-                if self._is_starting_or_live(tag):
-                    ipset = self.objects_by_id[tag]
-                    for ip in new_ips:
-                        ipset.add_member(ip, async=True)
-
-            self.endpoints_by_ep_id[endpoint_id] = endpoint
-            if old_prof_id and old_prof_id != new_prof_id:
-                ids = self.endpoint_ids_by_profile_id[old_prof_id]
-                ids.discard(endpoint_id)
-                if not ids:
-                    del self.endpoint_ids_by_profile_id[old_prof_id]
-            self.endpoint_ids_by_profile_id[new_prof_id].add(endpoint_id)
+        if new_prof_id != old_prof_id:
+            # Profile ID changed, or an add/delete.  the _xxx_profile_index
+            # methods ignore profile_id == None so we'll do the right thing.
+            _log.debug("Profile ID changed from %s to %s")
+            self._remove_profile_index(old_prof_id, endpoint_id)
+            self._add_profile_index(new_prof_id, endpoint_id)
+
+        # Since we've defaulted new/old_tags to set() if needed, we can
+        # use set operations to calculate the tag changes.
+        added_tags = new_tags - old_tags
+        unchanged_tags = new_tags & old_tags
+        removed_tags = old_tags - new_tags
+
+        # _extract_ips() will default old/new_ips to set() if there are no IPs.
+        old_ips = self._extract_ips(old_endpoint)
+        new_ips = self._extract_ips(endpoint)
+
+        # Remove *all* *old* IPs from removed tags.  For a deletion only this
+        # loop will fire, removed_tags will be all tags and old_ips will be
+        # all the old IPs.
+        for tag in removed_tags:
+            for ip in old_ips:
+                self._remove_mapping(tag, endpoint_id, ip)
+        # Change IPs in unchanged tags.
+        added_ips = new_ips - old_ips
+        removed_ips = old_ips - new_ips
+        for tag in unchanged_tags:
+            for ip in removed_ips:
+                self._remove_mapping(tag, endpoint_id, ip)
+            for ip in added_ips:
+                self._add_mapping(tag, endpoint_id, ip)
+        # Add *new* IPs to new tags.
+        for tag in added_tags:
+            for ip in new_ips:
+                self._add_mapping(tag, endpoint_id, ip)
 
         _log.info("Endpoint update complete")
 
+    def _add_mapping(self, tag_id, endpoint_id, ip_address):
+        """
+        Adds the given tag->endpoint->IP mapping to the index and updates
+        the ActiveIpset if present.
+
+        :return: True if the IP wasn't already in that tag.
+        """
+        ep_ids = self.ip_owners_by_tag[tag_id][ip_address]
+        ip_added = not bool(ep_ids)
+        ep_ids.add(endpoint_id)
+        self.ips_in_tag[tag_id].add(ip_address)
+        if ip_added and self._is_starting_or_live(tag_id):
+            _log.debug("Adding %s to active tag %s", ip_address, tag_id)
+            self.objects_by_id[tag_id].add_member(ip_address, async=True)
```

Comment:
I disagree with using add_member here. It seems to me that we could be making 1000 add_member calls because somebody adds a single tag to a profile. Better (more efficient and simpler) to make a single call to replace_member once we have got our data structures correct (which in fact might mean that add_member and remove_member can be retired, since I don't think they would then ever be called).